File: content-solutions/implementation/sim-rental-fee.js
function solution(records, fees) {
  const [baseTime, baseFee, unitTime, unitFee] = fees;
  const start = new Map();
  const total = new Map();
  for (const rec of records) {
    const [t, num, kind] = rec.split(" ");
    const [h, m] = t.split(":").map(Number);
    const minute = h * 60 + m;
    if (kind === "OUT") {
      start.set(num, minute);
      if (!total.has(num)) total.set(num, 0);
    } else {
      total.set(num, total.get(num) + minute - start.get(num));
      start.delete(num);
    }
  }
  for (const [num, s] of start) total.set(num, total.get(num) + 1439 - s);
  return [...total.keys()].sort().map((num) => {
    const t = total.get(num);
    if (t <= baseTime) return baseFee;
    return baseFee + Math.ceil((t - baseTime) / unitTime) * unitFee;
  });
}

File: content-solutions/implementation/sim-rental-fee.py
def solution(records, fees):
    base_time, base_fee, unit_time, unit_fee = fees
    start = {}
    total = {}
    for rec in records:
        t, num, kind = rec.split()
        h, m = map(int, t.split(":"))
        minute = h * 60 + m
        if kind == "OUT":
            start[num] = minute
            total.setdefault(num, 0)
        else:
            total[num] += minute - start.pop(num)
    for num, s in start.items():
        total[num] += 1439 - s
    answer = []
    for num in sorted(total):
        t = total[num]
        if t <= base_time:
            answer.append(base_fee)
        else:
            answer.append(base_fee + (t - base_time + unit_time - 1) // unit_time * unit_fee)
    return answer
